pkg/common/fflag/fflag.go

```diff
@@ -0,0 +1,121 @@
+// The fflag package implements a basic singleton pattern for the purpose of
+// providing SPIRE with a system-wide feature flagging facility. Feature flags
+// can be easily added here, in a single central location, and be consumed
+// throughout the codebase.
+package fflag
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+)
+
+// Flag represents a feature flag and its configuration name
+type Flag string
+
+// To add a feature flag, decleare it here along with its config name.
+// Then, add it to the `flags` package-level singleton map below, setting the
+// appropriate default value. Flags should always be opt-in and default to
+// false, with the only exception being flags that are in the process off being
+// deprecated.
+const (
+	// FlagForcedRotation controls whether or not the new APIs and
+	// extensions related to forced rotation and revocation are
+	// enabled or not. See #1934 for more information.
+	FlagForcedRotation Flag = "forced_rotation"
+
+	// FlagTestFlag is defined purely for testing purposes.
+	FlagTestFlag Flag = "i_am_a_test_flag"
+)
+
+var (
+	singleton = struct {
+		flags  map[Flag]bool
+		loaded *sync.Once
+		mtx    *sync.RWMutex
+	}{
+		flags: map[Flag]bool{
+			FlagForcedRotation: false,
+			FlagTestFlag:       false,
+		},
+		loaded: new(sync.Once),
+		mtx:    new(sync.RWMutex),
+	}
+)
+
+// Load initializes the fflag package and configures its feature flag state
+// based on the configuration input. Feature flags are designed to be
+// Write-Once-Read-Many, and as such, Load can be called only once. Load will
+// return an error if it is called more than once, if the configuration input
+// cannot be parsed, or if an unrecognized flag is set.
+func Load(rc RawConfig) error {
```

Comment:
It seems to me that we'll need distinct feature flags for server and agent (that sometimes overlap). Seems to me that this  common code should just take a RawConfig and the flag definitions  and return a "feature flag set" that can be queried to enable server and agent do specify their own....

Reply:
> It seems to me that we'll need distinct feature flags for server and agent (that sometimes overlap).

I agree that we will have distinct features in agent or server that sometimes overlap, but I don't think that means that the feature flag namespace must also be disjoint. It feels nice to have one place where they are all defined .. if we allow server/agent to declare their own, we'll end up with two or possibly three places that a flag can be defined.

Do you think it's a bad idea to share the namespace like this? Even if only server or agent ever consumes a given flag, I don't think it ends up being problematic?

Reply:
It just means that you can accidentally set a flag on one that was intended for the other. You won't get error diagnostics that you goofed up, it will silently succeed but have any effect.

Reply:
Yea, that's not ideal ... but I also wonder how often this will actually bite someone - if you're trying to enable a flag that is specific to an agent or server (presumably because you are testing something), you're likely to quickly figure out that you're not working with the one you need?

IMO it is worth the trade-off, but I'm not opposed to splitting it up (especially if it's not the trade-off that others would make). I'd have to put some thought into how to do that ... I feel like I'd lean in the "other" direction (i.e. have some package-level flags defined in server + agent packages, and provide them from cli call site when calling the common Load).

Reply:
Evan and I chatted a bit. He's going to make a small change to the Load function that takes the server/agent specific flags so we can get good diagnostics.